### PR TITLE
Ensure that ``Crystal.change_basis()`` returns a new object of the same type

### PR DESCRIPTION
**Pull request by dwpaley**
This is needed for the new cctbx.xfel.merge worker reindex_to_abc here: https://github.com/cctbx/cctbx_project/tree/reindex_to_abc

When we call change_basis on a MosaicCrystalSauter2014 (derived from Crystal), we need the result to be a transformed copy of the same class. Therefore the method is reimplemented in MosaicCrystalSauter2014 to use the correct copy constructor and duplicate the mosaic parameters.

### CODE REVIEW

**Comment by dwpaley** [comment]
This will close #5. Thanks @phyy-nx for flagging it.

**Comment by rjgildea** [comment]
@dwpaley I've updated this to add a `virtual clone()` method so that the original `Crystal.change_basis()` method will correctly copy the underlying `Crystal` object, even in derived classes, thus removing the need to have a near-duplicated copy of the function.
See also https://www.fluentcpp.com/2017/09/08/make-polymorphic-copy-modern-cpp/

File: newsfragments/420.feature [hunk 1]
```diff
@@ -0,0 +1,2 @@
+In class MosaicCrystalSauter2014, override function change_basis so that
+it returns a MosaicCrystalSauter2014 instead of the parent Crystal
```
Comment (rjgildea) [inline newsfragments/420.feature:2]:
```suggestion
Ensure that ``Crystal.change_basis()`` returns a new object of the same type
```

**Comment by dwpaley** [comment]
Amazing, thanks @rjgildea. I tried to find a cleaner way but I had no luck. This is great.

**Comment by dwpaley** [comment]
@ndevenish Is this ok to merge?

**Comment by ndevenish** [comment]
Yes, sorry, it dropped off my radar when I knew Richard was taking a look.

**Comment by dwpaley** [comment]
Not a problem, thanks!

**Pull request merged**